resources/netlog_defs.ts

```diff
@@ -1,53 +1,68 @@
 import { PluginCombatantState } from '../types/event';
 import { NetFieldsReverse } from '../types/net_fields';
+import { NetParams } from '../types/net_props';
 
+type AnalysisFilter = 'all' | 'filter' | 'none' | 'never';
 export type LogDefinition = {
-  // The log id, as a decimal string, minimum two characters.
-  type: string;
-  // The informal name of this log (must match the key that the LogDefinition is a value for).
-  name: string;
-  // The plugin that generates this log.
-  source: 'FFXIV_ACT_Plugin' | 'OverlayPlugin';
-  // Parsed ACT log line type.  OverlayPlugin lines use the `type` as a string.
-  messageType: string;
-  // If true, always include this line when splitting logs (e.g. FFXIV plugin version).
-  globalInclude?: boolean;
-  // If true, always include the last instance of this line when splitting logs (e.g. ChangeZone).
-  lastInclude?: boolean;
-  // True if the line can be anonymized (i.e. removing player ids and names).
-  canAnonymize?: boolean;
-  // If true, this log has not been seen before and needs more information.
-  isUnknown?: boolean;
-  // Fields at this index and beyond are cleared, when anonymizing.
-  firstUnknownField?: number;
-  // A map of all of the fields, unique field name to field index.
-  fields?: { [fieldName: string]: number };
-  // A list of field ids that *may* contain RSV keys (for decoding)
-  possibleRsvFields?: readonly number[];
-  subFields?: {
-    [fieldName: string]: {
-      [fieldValue: string]: {
-        name: string;
-        canAnonymize: boolean;
+  [K in keyof typeof latestLogDefinitions]: {
```

Comment:
I have some minor concerns about this type change, insomuch as this changes the type of `LogDefinition` from a single possible constructed type to a union of every possible log line type defined in `latestLogDefinitions`. This won't have any runtime performance issues, but it does increase the compile time for functionally every webpack package as it has to construct those types.

Type enforcement is already being done on the `latestLogDefinitions` type down below the definition, where there's a `console.assert`.

This might be considered "premature optimization" but given the already long compile times and webpack dev server memory usage etc it's something to consider at least.

To be clear, I'm not suggesting removing this change or anything, I just want to facilitate discussion on the point.

Reply:
It's a fair point.  Like you say, it's really only added load during compile time (and I doubt there's much if anyone using webpack outside of those of us doing active development). But I think there might be a lway to lower the cost & achieve the same outcome.

The objective was to:
- ensure `name` matches the key (which it needs to and we say it should but weren't enforcing)
 - ensure `analysisFilter` contains a valid and matching `NetParams` object or array.
 - ensure that use of `analysisFilter` and `filterCombatantIdFields` is appropriate for the `includeForAnalysis` type

I think we could still accomplish that by switching `LogDefinition` back to its original constructed type (genericizing the typing on the `name` and `analysisFilter` props), and then merging the original `console.assert()` check with the second one I added into a single assert, e.g.:

```typescript
type LogDefinitionStrict = {
  [K in LogDefinitionTypes]:
  Omit<LogDefinition, 'name' | 'includeForAnalysis' | 'analysisFilter' | 'filterCombatantIdFields'>
  & {
    name: K;
  }
  & (typeof latestLogDefinitions[K]['includeForAnalysis'] extends 'filter'
    ? {
      includeForAnalysis: 'filter';
      analysisFilter: NetParams[K] | readonly NetParams[K][];
      filterCombatantIdFields?: number | readonly number[];
    }
    : typeof latestLogDefinitions[K]['includeForAnalysis'] extends 'all' ? {
        includeForAnalysis: 'all';
        filterCombatantIdFields?: number | readonly number[];
        analysisFilter?: undefined;
      }
    : typeof latestLogDefinitions[K]['includeForAnalysis'] extends 'none' | 'never' ? {
        includeForAnalysis: 'none' | 'never';
        analysisFilter?: undefined;
        filterCombatantIdFields?: undefined;
      }
    : never);
};
const logDefinitionStrict: LogDefinitionStrict = latestLogDefinitions;
console.assert(logDefinitionStrict);
```

I think that would eliminate at least some of the compile cost? I'm honestly not sure though.

Reply:
If the goal is to enforce the typing of `name` that could be accomplished with this, for example:
```ts
export type LogDefinition = {
// ...
  // The informal name of this log (must match the key that the LogDefinition is a value for).
  name: keyof typeof latestLogDefinitions;
```

If the goal is for analysis information to have stricter typing, it could be its own sub-property, something like:

```ts
export type LogDefinition = {
// ...
  // logic: Treat `undefined` here as `includeForAnalysis: 'none'`
  analysisOptions?: AnalysisOptionsType<K>;
};

type AnalysisFilter = 'all' | 'filter' | 'none' | 'never';
export type AnalysisOptionsType<K extends keyof typeof latestLogDefinitions> = {
  includeForAnalysis: Extract<AnalysisFilter, 'none' | 'never'>;
} | {
  includeForAnalysis: Extract<AnalysisFilter, 'filter'>;
  analysisFilter: NetParams[K] | readonly NetParams[K][];
  filterCombatantIdFields?: number | readonly number[];
} | {
  includeForAnalysis: Extract<AnalysisFilter, 'all'>;
  filterCombatantIdFields?: number | readonly number[];
};

export type LogDefinitionMap = { [name in keyof typeof latestLogDefinitions]: LogDefinition<name> };
```

I believe that this should allow type erasure to ensure that there is only one derived type for `LogDefinition`, though I'm not sure how the downstream `anonymizer.ts` and `splitter.ts` files would need to be updated to include the type properly.

This has the added benefit of ensuring typing is correct for the include vs exclude properties. Type narrowing can be used to ensure the correct derived `AnalysisOptionsType` is determined, e.g.
```ts
      if (def.includeForAnalysis === 'filter') {
        // type narrowing means that `def.analysisFilter !== undefined` is not required here as it can never be `undefined` if the `includeForAnalysis` property is `filter`
```

Reply:
>   name: keyof typeof latestLogDefinitions;

Still thinking about the rest of the suggestion, but I think this will only limit 'name' to any log definition type; it wouldn't ensure that the value matches the key in latestLogDefinitions?

Reply:
Sorry, I had left out the generic parameters in my previous comment about the `export type LogDefinition` line, that's included below along with the "fix" for allowing any set:

```ts
export type LogDefinition<K extends keyof typeof latestLogDefinitions> = {
  // The log id, as a decimal string, minimum two characters.
  type: (typeof latestLogDefinitions)[K]['type'];
  // The informal name of this log (must match the key that the LogDefinition is a value for).
  name: K;
```

(this could be repeated for any other non-optional property)

(also probably every instance of `typeof latestLogDefinitions` could be replaced with `LogDefinitions` in the file, that would be a bit easier to read)